### PR TITLE
[ELY-2544] [ELY-2254] Provide a LoginModule compatible security realm.

### PR DESCRIPTION
**Pull request by skyllarr**
issue: https://issues.redhat.com/browse/ELY-2544

This issue is resolved on maintenance branch by backporting the issue [ELY-2254](https://issues.redhat.com/browse/ELY-2254) to the maintenance branch.
Upstream branch does not need any changes.

This PR is only a backporting of [ELY-2254](https://issues.redhat.com/browse/ELY-2254)

Upstream PR: https://github.com/wildfly-security/wildfly-elytron/pull/1623

### CODE REVIEW

**Review by fjuma — APPROVED** [review]

**Comment by yersan** [comment]
It looks like PR links are wrong: PR title and commit message are related to **ELY-2254**, however, the description is linked to [**ELY-2544**](https://issues.redhat.com/browse/ELY-2544)

**Comment by skyllarr** [comment]
Hello @yersan , this issue (ELY-2254) is meant to backport the changes of ELY-2544 to the 1.15.x maintanence branch. So the PR links are correct here

**Comment by yersan** [comment]
Hi @Skyllarr , sorry, I guess the title and Jira link in the description confused me. 

> this issue (ELY-2254) is meant to backport the changes of ELY-2544 to the 1.15.x maintanence branch. So the PR links are correct here.

I am not familiar with how Elytron manages its issues. On WildFly / WildFly Core we follow the rule where each PR commit, PR title, and PR Jira link match all together for each PR. Hence my confusion with this one, where the commit message and PR title use ELY-2254 but the Jira link is ELY-2544.

You mentioned this issue is ELY-2254, however, that issue is already resolved and since ELY-2544 Jira is still open, I understand that this PR is for ELY-2544. Ideally, ELY-2544 should be in the pull request sent with a link to this PR. Don't get me wrong, I am not asking to change anything now, just wanted to mention that it could confuse people outside of Elytron who are used to working on WildFly (like me :)). In any case, now I have a better picture of Jira links on EAP7-2068.

**Comment by skyllarr** [comment]
Hi @yersan , I understand the confusion and I'm sorry about that. You're right, this issue was not linked in the ELY-2544 jira issue which was my mistake, it is now linked. 

It is confusing because the ELY-2544 issue is part of a maintenance RFE where the only change needed in elytron was to backport the ELY-2254 to the maintenance branch (similar issue numbers do not help either :D ) And since this is a backport, the commit title remained ELY-2254. The upstream branch specifically does not need any elytron changes for the EAP7-2068 RFE.

I have now added both issue numbers to the PR title to make it more clear, and will also add the explanation to the Jira issue description. Thanks a lot for looking into this!

File: auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java [hunk 1]
```diff
@@ -60,28 +62,74 @@
...
 
     /**
      * Construct a new instance.
      *
-     * @param loginConfiguration the login configuration name to use
+     * @param entry              JAAS configuration file entry (must not be {@code null})
+     * @param jaasConfigFilePath path to JAAS configuration file
+     * @param classLoader        classLoader to use with LoginContext, this class loader must contain LoginModule CallbackHandler classes
      */
-    public JaasSecurityRealm(final String loginConfiguration) {
```
Comment (fjuma) [inline auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java:72]: Just a note for other reviewers that `JaasSecurityRealm` is private API.

File: auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java [hunk 2]
```diff
@@ -60,28 +62,74 @@
...
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param entry       JAAS configuration file entry (must not be {@code null})
+     * @param classLoader classLoader to use with LoginContext, this class loader must contain LoginModule CallbackHandler classes
+     */
+    public JaasSecurityRealm(final String entry, final ClassLoader classLoader) {
+        this(entry, null, classLoader);
```
Comment (fjuma) [inline auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java:87]: Just to check, can `classLoader` be null here?

File: auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java [hunk 3]
```diff
@@ -60,28 +62,74 @@
...
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param entry       JAAS configuration file entry (must not be {@code null})
+     * @param jaasConfigFilePath path to JAAS configuration file
+     */
+    public JaasSecurityRealm(final String entry, final String jaasConfigFilePath) {
+        this(entry, jaasConfigFilePath, null);
```
Comment (fjuma) [inline auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java:97]: Just to check, can `jaasConfigFilePath` be null here?

File: auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java [hunk 4]
```diff
@@ -103,47 +151,58 @@ public SupportLevel getCredentialAcquireSupport(final Class<? extends Credential
     @Override
     public SupportLevel getEvidenceVerifySupport(final Class<? extends Evidence> evidenceType, final String algorithmName) throws RealmUnavailableException {
         Assert.checkNotNullParam("evidenceType", evidenceType);
-        return PasswordGuessEvidence.class.isAssignableFrom(evidenceType) ? SupportLevel.SUPPORTED : SupportLevel.UNSUPPORTED;
+        return SupportLevel.POSSIBLY_SUPPORTED;
```
Comment (fjuma) [inline auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java:154]: Just curious, why does this need to be changed?

File: auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java [hunk 5]
```diff
@@ -103,47 +151,58 @@ public SupportLevel getCredentialAcquireSupport(final Class<? extends Credential
...
                 final Method setSecurityInfo = handler.getClass().getMethod("setSecurityInfo", Principal.class, Object.class);
                 setSecurityInfo.invoke(callbackHandler, principal, evidence);
                 return callbackHandler;
-            } catch (Exception e) {
-                throw ElytronMessages.log.failedToInstantiateCustomHandler(e);
+            } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+                // ignore if this method does not exist
+                return handler;
             }
+        } else if (Security.getProperty("auth.login.defaultCallbackHandler") != null) {
```
Comment (fjuma) [inline auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java:194]: Would be good to create and make use of a constant for "auth.login.defaultCallbackHandler".

File: auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java [hunk 6]
```diff
@@ -183,31 +246,32 @@ public SupportLevel getEvidenceVerifySupport(final Class<? extends Evidence> evi
...
-                return successfulLogin;
-            } else {
-                return false;
+            } finally {
+                Thread.currentThread().setContextClassLoader(oldClassLoader);
             }
+            return successfulLogin;
         }
 
-        public boolean exists() throws RealmUnavailableException {
```
Comment (fjuma) [inline auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java:210]: Just to check, why is this being changed?

**Comment by fjuma** [comment]
@Skyllarr I just realized that this PR just backports ELY-2254, there's nothing new being introduced for ELY-2544. I'll mark my previous comments as resolved.

**Review by fjuma — APPROVED** [review]

**Review by darranl — APPROVED** [review]
Approving as a backport, if there is anything specific to the backport we have missed let us know otherwise looks good.

**Pull request merged**